<a href="https://colab.research.google.com/github/fedhere/FDSFE2024_Final/blob/main/Instructions_FDSFE_2024_FINAL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# FINAL EXAM: Foundations of Data Science for Everyone 2024

**GOAL:** Use Machine Learning methods to predict the presence of heart disease in a patient.


The final will be graded 0-100.

- There are 100 available points from 10 tasks described below.

-  There are additional optional tasks that can earn you additional points (and compensate for points lots in the required tasks).

- Up to 15 additional points can be awarded for a clean and tidy deliver. This includes: properly formatted print statements, control of decimal digits in the answers (should generally be <=2), and a code that can be run from top to bottom without stopping.

- In many cases, doing well on the tasks that you do even if you do not finish all the tasks will result in a better grade!


- If you are concerned that the exam pressure did not allow you to perform well, at the end of the exam schedule notify me and Willow that you want to schedule a 1-1 interview to explain what you did in the final and what you should have done instead. If possible this request will be accommodated and you will be able to recover some of the points in the exam. Note however that this can go two ways: if you did things right in the exam and you cannot explain them you will lose points and your interview may result in lowering your grade. Only use this option if you feel you underperformed because of exam pressure.


The data is a modified version of this dataset https://archive.ics.uci.edu/dataset/45/heart+disease that includes fewer variables and some values have been deliberately modified.

The goal is to predict heart desease based on medical and personal data from patients.

# TABLE OF CONTENT
- TASK 1 Acquire in the data _10 points_
  - Get the data from the folder "data" inside of the exam repository (2 pts)
  - Show the top and the bottom 10 rows of the data. (2 pts)
  - Answer the question: why do I ask to show both the top and the bottom of the dataset? (4 pts)
  - Report the number of rows and columns in the dataset in a print statement that reads: "The number of rows in the dataset is: N; the number of columns is M" with N and M set to the appriopriate number of course. (2 pts)

- TASK 2 Read carefully the data dictionary  _5 points_

- TASK 3 Check the number of values in the target variable _10 points_
  - print the values `target` can take (2 pts)
  - Answer the question: Based on the  number of values that the target variable can take, what is the machine learning task we are engaging in if we want to predict target? (4 pts)
  - Print the number of instances for each value of `target` (2 pts)
  - OPTIONAL: Choose a way to visualize the number of instances in each value in `target`
  - Answer the question: why is it important to do check the number of instances of each value? can you give an example of a pair of numbers (instances of one value e.g. =0, instances of the other e.g. =1) that would cause problem when making a prediction?(4 pts)


- TASK 4 Address missing values _15 points_
  - Answer the question: are there missing values?  (write code to answer the question and explain what in the output of the code shows that there are missing values. There are many ways to do this). (2 pts)
  - Visualize missing values (we did this in class in at least two ways) (4 pts)
  - Decide what to do with missing values for each column or row (or cell) and "fix" the missing values as you see fit (2 pts)
  - Describe what you did and JUSTIFY your choice of way to address the problem (4 pts)
  - Check that there are no more missing values (1 pts)
  - Report on the _percentage_ of data that was loss in the process of handling missing values: how many rows, how many columns, how many entries total. (2 pts)


- TASK 5 identify the numerical and categorical variables _6 points_
  - save the name of each binary column and print it (2 pts)
  - save the name of each categorical column  and print it (2 pts)
  - save the names of each numerical column (for example in a list) and print it (2 pts)
  

- TASK 6 explore the correlation of the features _14 points_
  - Visualize the correlation of  features (including the target) and describe your plot in a caption that includes the answers to the following questions:
    - which pair of variables are most strongly correlated with each other? (2 pts)
    - which is the most strongly correlated variable with the target variable? (2 pts)
    - which is the most strongly correlated _numerical_ variable with the target variable? (2 pts)
  - plot the distribution of the most correlated variable with target (and describe the plot and the insight that you can derive from it in a caption) (2 pts)
  - plot the distribution of the most correlated _numerical_ variable with target (and describe the plot and the insight that you can derive from it in a caption) (2 pts)
  - make a pair-plot or scatter-matrix plot of all variables and describe what you see including how the plot is designed, what is the content of each subplot on the _diagonal axis_ and on the _off diagonal axis_ (note, this takes a few minutes) (4 pts)


- TASK 7 Model choice _5 points_
  - Answer the question: why is logistic regression a reasonable model to use if I want to work in a univariate framework? (3 pts)
  - Answer the question: why is random forest a reasonable model to use for the prediction overall? (5 pts)


- OPTIONAL TASK Perform a Univariate Logistic regression using the maximally corrleated numerical variable with target as the exhogenous variable
  - prepare the dataset for training and testing
  - report what is the size of the training and of the testing set. Why did you choose that size split?
  - Answer the question: why is it important to evaluate the performance on both a training and a testing set?
  - perform the logistic regression
  - report the model accuracy (i.e. score) on training and testing datasets
  - comment on the accuracy and the risk of overfitting
  - use the given code to print out a confusion matrix for your model. Write a caption for it  

- TASK 8 model target using a random forest model _15 points_
  - generate the feature set `X` to include all variables in the problem and `y` to `target` (1 pts)
  - set the model with the following hyperparameters: `n_estimators=100, max_depth=15, random_state=302` (1 pts)
  - prepare the dataset for training and testing (2 pts)
  - Answer the question (if you did not in the previous task): why is it important to evaluate the performance on both a training and a testing set? (5 pts)
  - Answer the question: why is scaling or standardizing the data not necessary with this model (note, it was also not necessary with the logistic regression model, but for different reasons!) (5 pts)
  - run a random forest model (1 pts)

- TASK 9 model evaluation _10 points_
  - report the model accuracy (i.e. score) on training and testing datasets (1 pts)
  - comment on the accuracy and overfitting (5 pts)
  - use the given code to print out a confusion matrix for your model. Write a caption for it (4 pts)

- TASK 10 fix overfitting _10 points_

  - What can you do about overfitting? control the appropriate hyperparameter(s) and redo the model with those values, report the training and testing score (what do they say about overfitting?) and redo the confusion matrix (5 pts)
  - Answer the question: Is the model better? describe how and why (or why not) (5 pts)


- OPTIOINAL EXTRA CREDIT : feature importance



In [ ]:
#system imports
import sys

#useful package for data handling
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

#plotting packages
import matplotlib # collection of functions for scientific and publication-ready visualization
import matplotlib.pyplot as plt
import seaborn as sns

#missing variables
import missingno as msno
from sklearn.impute import KNNImputer

#model preprocessing
from sklearn import preprocessing
from sklearn.model_selection import train_test_split

#modeling packages
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

#for model evaluation
from sklearn.metrics import confusion_matrix


# TASK 1 Acquire in the data _10 points_
- Get the data from the folder "data" inside of the exam repository
- Show the top and the bottom 10 rows of the data.
- Answer the question: why do I ask to show both the top and the bottom of the dataset?
- Report the number of rows and columns in the dataset in a print statement that reads: "The number of rows in the dataset is: N; the number of columns is M" with N and M set to the appriopriate number of course.


In [ ]:
heart = ...

In [ ]:
...

In [ ]:
# check below: the bottom of the file should look like this if you read it in correctly
... #your code here

age  gender   cp  trtbps   chol  fbs  restecg  thalachh  exng  oldpeak  \
293   67     1.0  2.0     152  212.0  0.0      0.0       150   0.0      0.8   
294   44     1.0  0.0     120  169.0  0.0      1.0       144   1.0      NaN   
295   63     1.0  0.0     140  187.0  NaN      0.0       144   1.0      4.0   
296   63     0.0  0.0     124  197.0  NaN      1.0       136   1.0      0.0   
297   59     1.0  0.0     164  176.0  1.0      0.0        90   0.0      1.0   
298   57     0.0  0.0     140  241.0  0.0      1.0       123   1.0      0.2   
299   45     1.0  3.0     110  264.0  0.0      1.0       132   0.0      1.2   
300   68     1.0  0.0     144  193.0  1.0      1.0       141   0.0      3.4   
301   57     1.0  0.0     130  131.0  0.0      1.0       115   1.0      NaN   
302   57     0.0  NaN     130  236.0  0.0      0.0       174   0.0      0.0   

     slp  caa  thall  target  
293  1.0    0      3       0  
294  0.0    0      1       0  
295  2.0    2      3       0  
296  1.0    0      2       0  
297  1.0    2      1       0  
298  1.0    0      3       0  
299  1.0    0      3       0  
300  1.0    2      3       0  
301  1.0    1      3       0  
302  1.0    1      2       0

Why do I ask to show both the top and the bottom of the dataset?  

_YOUR ANSWER HERE_


In [ ]:
print("...

The number of rows in the dataset is: 303; the number of columns is 14


# TASK 2 Read carefully the data dictionary that follows; it will help you in the analysis! _5 points_

Columns name description and units (from UCI data website):
- `cp`: chest pain type
      Value 0: typical angina (most serious)
      Value 1: atypical angina
      Value 2: non-anginal pain
      Value 3: asymptomatic (least serious)
- `trestbps`: resting blood pressure (in mm Hg on admission to the hospital)
- `chol`: serum cholestoral in mg/dl
- `fbs`: fasting blood sugar > 120 mg/dl (1 = true; 0 = false)

      A fasting blood sugar level less than 100 mg/dL is normal. From 100 to 120 mg is considered prediabetes. If it is 125 mg/dL or higher on two separate tests, you have diabetes.
- `restecg`: resting electrocardiographic results
- `thalach`: maximum heart rate achieved [beats per minute]
- `exang`: exercise induced angina (1 = yes; 0 = no)
- `oldpeak`: ST depression induced by exercise relative to rest.
      ST depression is a finding on an electrocardiogram (EKG) that indicates an underlying heart condition.
- `slope`: the slope of the peak exercise ST segment
- `ca`: number of major vessels (0-3) colored by flouroscopy
- `thal`: 3 = normal; 6 = fixed defect; 7 = reversable defect.

      This refers to Thalassemia, an inherited blood disorder that causes your body to have less hemoglobin than normal


- `target`: refers to the presence of heart disease in the patient (1=yes, 0=no)
      Value 0: < 50% diameter narrowing
      Value 1: > 50% diameter narrowing

WRITE YOUR NAME HERE  BELOW THIS LINE IN THIS CELL TO DEMONSTRATE YOU READ THIS LIST

# TASK 3 Check the number of values in the target variable _10 points_
- print how many values `target` can take
- Answer the question: Based on the  number of values that the target variable can take, what is the machine learning task we are engaging in if we want to predict target?
- Print the number of instances for each value of `target`.
- OPTIONAL: Choose a way to visualize the number of instances in each value in `target`
- Answer the question: why is it important to do check the number of instances of each value? can you give an example of a pair of numbers (instances of one value, instances of the other, e.g. =0, instances of the other e.g. =1) that would cause problem when making a prediction?



In [ ]:
...

Based on the number of values that the target variable can take, what is the machine learning task we are engaging in if we want to predict target?

_YOUR ANSWER HERE_


In [ ]:
...

In [ ]:
...

why is it important to do check the number of instances of each value? can you give an example of a pair of numbers (instances of one value, instances of the other, e.g. =0, instances of the other e.g. =1) that would cause problem when making a prediction?

_YOUR ANSWER HERE_


# TASK 4 Address missing values _12 points_
- Are there missing values?  (write code to answer the question and explain what in the output of the code shows that there are missing values. There are many ways to do this).
- Visualize missing values (we did this in class in at least two ways) and remember that when you make a plot you also need to have a caption that describes how the plot is designed and what insight you derive from it for it to ount!
- Decide what to do with missing values for each column or row (or cell)
- Describe what you did and JUSTIFY your choice of way to address the problem
- Check that there are no more missing values
- Report on the _percentage_ of data that was loss in the process of handling missing values: how many rows, how many columns, how many entries total.

In [ ]:
...

In [ ]:
...

In [ ]:
# Missing data visualization and caption
...

In [ ]:
...

In [ ]:
...

In [ ]:
...

Describe what you did and JUSTIFY your choice of way to address the problem

_ YOUR ANSWER HERE_

In [ ]:
...

In [6]:
print("Amount of data lost: {0}% observations, {1}% variables".format("X", "Y"))
print("Amount of data lost: {0}% entries".format("Z"))

Amount of data lost: X% observations, Y% variables
Amount of data lost: Z% entries


# TASK 5 identify the numerical and categorical variables _6 points_
- save the names of each numerical column (for example in a list) and print it
- save the name of each binary column and print it
- save the name of each categorical column  and print it

(hint: since there are missing values, some columns will be read as float64 just because they include NaNs even if the are really intergers. You can check what variables have exactly 2 values, and what variables have <10 values, and assume that this is a good indicator of the categorical nature of a column/feature)

In [ ]:
qualitative = []
quantitative = []
binary = []
....

In [ ]:
...

In [ ]:
...

In [ ]:
...

# TASK 6 explore the correlation of the features _14 points_
- Visualize the correlation of  features (including the target)
- Describe your plot in a caption that includes the answers to the following questions:
  - which pair of variables are most strongly correlated with each other?
  - which is the most strongly correlated variable with the target variable?
  - which is the most strongly correlated _numerical_ variable with the target variable?
  - which is the least strongly correlated variable with the target variable? (you will use this later so take care this task is done well!)
- plot the distribution of the most correlated variable with target (and describe the plot and the insight that you can derive from it in a caption)
- plot the distribution of the most correlated _numerical_ variable with target (and describe the plot and the insight that you can derive from it in a caption)
- make a pair-plot or scatter-matrix plot of all variables and describe what you see including how the plot is designed, what is the content of each subplot on the _diagonal axis_ and on the _off diagonal axis_ (note, this takes a few minutes)

(hint: think carefully about this: you want to find out which variables may be  good predictors of the target. if the variable is linearly correlated, which is what you are probing here, i.e. if plotting the variable against the target results in a line, _regardless of the specific slope of that line_, that would be a good predictor!)


In [ ]:
X = ...
Y = ...
...

In [ ]:
...

In [ ]:
...

In [ ]:
print("maximum correlation overall: {} and {}".format("...", "..."))

In [ ]:
...

In [ ]:
...

In [ ]:
...

In [ ]:
print(f"Maximum correlation of targets is ...

In [ ]:
print(f"Maximum correlation of targets with numerical variables is ...

In [ ]:
print("max correlation with target {0:};  max correlation for numerical variables with target {1:}".format(maxcorr, maxcorrnum))

max correlation with target exng;  max correlation for numerical variables with target thalachh


In [ ]:
# plot the distribution of the most correlated  variable in a plot and comment in a caption
...

In [ ]:
# plot the distribution of the most correlated numerical variable in a plot and comment in a caption
...

In [ ]:
# pair plot or scatter matrix plot and caption
...


# TASK 7 Model choice _5 points_

You will make two models:
- a univariate logistic regression using the single most correlated numerical variable to predict target
- a random forest using all variables to predict target

For you to do:
- Answer the question: why is logistic regression a reasonable model to use if I want to work in a univariate framework?
- Answer the question: why is random forest a reasonable model to use for the prediction overall?

Why is logistic regression a reasonable model to use if I want to work in a univariate framework?


_YOUR ANSWERS HERE_

Why is random forest a reasonable model to use for the prediction overall?


_YOUR ANSWERS HERE_


# OPTIONAL TASK Perform a Univariate Logistic regression using the maximally corrleated numerical variable with target as the exhogenous variable
apprioriately split the data into training and testing
- set `X` to the numerical column with the highest correlation with target, as identified in the previous step. Set y to the target column.
- prepare the dataset for training and testing
- report what is the size of the training and of the testing set. Why did you choose that size split?
- Answer the question: why is it important to evaluate the performance on both a training and a testing set?
- perform the logistic regression
- report the model accuracy (i.e. score) on training and testing datasets.
- comment on the accuracy and the risk of overfitting
- use the given code to print out a confusion matrix for your model. Write a caption for it (note: the code assumes the following:
    - the test target variable is called `y_test`
    - the model predictions on the test data are stored in the variable `Y_pred`

In [ ]:
X = ...
y = ...
...

In [ ]:
...

In [10]:
model = ...
print("Training Score: ...")
print("Testing Score: ...")



Training Score: ...
Testing Score: ...


Why is it important to evaluate the performance on both a training and a testing set?

_YOUR ANSWER HERE_

Comment on the accuracy and the risk of overfitting (please **don't** say "accuracy is good" or "accuracy is bad". Rather explain what the score value you printed above actually means in terms of predictions)

_YOUR ANSWER HERE_


In [ ]:
# CONFUSION MATRIX CODE

LR_confusion_matrix = confusion_matrix(y_test,Y_pred)
class_names = ["no disease", "heart disease"]
fig,ax = plt.subplots()
tick_marks = np.arange(len(class_names))

sns.heatmap(pd.DataFrame(LR_confusion_matrix)
            , annot = True, cmap = 'Greens', fmt = 'g')
ax.xaxis.set_label_position('top')
plt.xticks(tick_marks + 0.5, class_names)
plt.yticks(tick_marks + 0.5, class_names)
plt.tight_layout()
plt.title('Confusion matrix for logistic regression')
plt.ylabel('Actual label')
plt.xlabel('Predicted label')
plt.show()

# TASK 8 model target using a random forest model _15 points_
  - generate the feature set `X` to include all variables in the problem and `y` to `target`
  - set the model with the following hyperparameters: `n_estimators=100, max_depth=15, random_state=302`
  - prepare the dataset for training and testing
  - Answer the question (if you did not in the previous task): why is it important to evaluate the performance on both a training and a testing set?
  - Answer the question: why is scaling or standardizing the data not necessary with this model (note, it was also not necessary with the logistic regression model, but for different reasons!)
  - run a random forest model



In [ ]:
X = ...
y = ...

...

In [ ]:
...

In [ ]:
model = ...



Answer the question (if you did not in the previous task): Why is it important to evaluate the performance on both a training and a testing set?

_YOUR ANSWER HERE_
  

Why is scaling or standardizing the data not necessary with this model (note, it was also not necessary with the logistic regression model, but for different reasons!)

_YOUR ANSWER HERE_

# TASK 9 model evaluation _10 points_
  - report the model accuracy (i.e. score) on training and testing datasets
  - comment on the accuracy and overfitting
  - use the given code to print out a confusion matrix for your model. Write a caption for it

In [11]:

...
print("Training Score: ...")
print("Testing Score: ...")




Training Score: ...
Testing Score: ...


Comment on the accuracy and the risk of overfitting (please **don't** say "accuracy is good" or "accuracy is bad". Rather explain what the score value you printed above actually means in terms of predictions)

_YOUR ANSWER HERE_


In [ ]:
# CONFUSION MATRIX CODE
RF_confusion_matrix = confusion_matrix(y_test,Y_pred)
class_names = ["no disease", "heart disease"]
fig,ax = plt.subplots()
tick_marks = np.arange(len(class_names))

sns.heatmap(pd.DataFrame(RF_confusion_matrix), annot = True, cmap = 'Greens', fmt = 'g')
ax.xaxis.set_label_position('top')
plt.xticks(tick_marks + 0.5, class_names)
plt.yticks(tick_marks + 0.5, class_names)
ax.xaxis.set_label_position('top')
plt.tight_layout()
plt.title('Confusion matrix for logistic regression')
plt.ylabel('Actual label')
plt.xlabel('Predicted label')
plt.show()

# TASK 10 fix overfitting _10 points_
What can you do about overfitting? control the appropriate hyperparameter(s) and redo the model with those values

Answer the question: Is the model better? describe how and why (or why not)



In [12]:

model = ...
print("Training Score: ...")
print("Testing Score: ...")



Training Score: ...
Testing Score: ...


In [ ]:
# CONFUSION MATRIX CODE
RF_confusion_matrix = confusion_matrix(y_test,Y_pred)
class_names = ["no disease", "heart disease"]
fig,ax = plt.subplots()
tick_marks = np.arange(len(class_names))

sns.heatmap(pd.DataFrame(RF_confusion_matrix), annot = True, cmap = 'Greens', fmt = 'g')
ax.xaxis.set_label_position('top')
plt.xticks(tick_marks + 0.5, class_names)
plt.yticks(tick_marks + 0.5, class_names)
ax.xaxis.set_label_position('top')
plt.tight_layout()
plt.title('Confusion matrix for random forest controlling overfit regression')
plt.ylabel('Actual label')
plt.xlabel('Predicted label')
plt.show()

Is the model better? describe how and why (or why not)

_YOUR CODE HERE_

# EXTRA CREDIT

- perform feature importance analysis and intepret it in the light of the correlation analysis
- what can you conclulde about the feature importance keeping into account the correlation analysis you did before?

Thanks [towards data science](https://towardsdatascience.com/a-summary-of-the-basic-machine-learning-models-e0a65627ecbe)